### DISCUSSION TITLE
Concatenate adjacent strings

### DISCUSSION THREAD

**Discussion by sirinath** [dotnet/csharplang#5604] [Language Ideas]
If adjacent string be concatenated without `+` which would  be more neater and  easy when splitting string into multiple lines.

```csharp
string s = "..."
           "..."
           "...";
```

**Comment by DavidArno** [comment]
Sorry, but I disagree. It just looks wrong, so I don't agree it looks neater. And how is typing `+` hard and thus how is not typing it measurably easier?

**Reply by bernd5** [reply]
In C and C++ this syntax is allowed and quite common - but I agree and think `+` is way more obvious.

**Comment by CyrusNajmabadi** [comment]
> and easy when splitting string into multiple lines.

It's already pretty easy to split strings into multiple lines. Just have your caret in the string and press enter. We take care of it for you :)

As for 'neater', I think it's reasonable to argue that. However, to me, it feels like a minor savings.

**Comment by theunrepentantgeek** [comment]
One of the key reasons that GUIs dominated over command line interfaces is this:

_Recognition_ is faster and easier that _remembering_.

There's been a whole bunch of cognitive research that drills into why - it makes for fascination reading. 

I'd suggested that _recognition_ of the + operator as concatenation will have lower cognitive overhead than requiring readers to _remember_ that the absence of an operator means concatenation. 

Or, stated otherwise, leaving out the operator is very likely harder to read (slower comprehension) than keeping it in, even for those familiar with the syntax.

**Reply by sirinath** [reply]
If the string contains operators (JS, SQL, etc.) it looks messy. 

If one prefers one can use the +.

**Reply by HaloFour** [reply]
C# has verbatim string literals that can span multiple lines, and is adding raw strings which can also span multiple lines.

https://github.com/dotnet/csharplang/issues/4304

**Reply by sirinath** [reply]
I know.

You loose control of new lines in the string when you use them.

**Comment by sirinath** [comment]
What this does is:
- break or wrap string into multiple lines without introducing new lines
- ability to individually control whitespace for each line if needed
- reduce noise by not having `+` when this might be an distraction
- be close to treatment of string in other languages of the same family

**Reply by HaloFour** [reply]
`+` isn't noise, it's intention.

**Reply by sirinath** [reply]
That depends on the situation (content of the string) as I mentioned before. So it is a convenience if one can omit it. This does not mean `+` will be unavailable. In situation this make sense this can be used.

**Reply by CyrusNajmabadi** [reply]
Yes. This could be done. The question is should it. It's extremely minor in savings and value. 

Having used this in other languages I genuinely find it harder to read. 

There's already a good way to write this today which is idiomatic and simple. It's going to be a high bar to clear to add a new form with such a minimal improvement

**Reply by HaloFour** [reply]
The context is that you want to concatenate strings.  The `+` is the operator to do so.  I don't like the idea that by simply having two strings near each other in terms of whitespace will result in their concatenation.

You mention this feature in the context of embedded literals of JavaScript and SQL, both of which are motivating use cases for the inclusion of raw strings.  Both of which are usually formatted with newlines included, as whitespace is generally needed to separate whatever syntax would appear on the previous line from what will appear on the next line anyway, and because you'd want to see those new lines in the resultant string.

**Reply by sirinath** [reply]
JS, SQL is one.

Another big motivator is a list of items in a string. Instead of going horizontal for a long list it is more convenient to go vertical. In this case new lines and white space matter. This is not a big change or difficult to do but more convenient in:
- code in string though other options exist
- lists of items in string where not satisfactory option exist

**Reply by spydacarnage** [reply]
So, you're suggesting that 
```cs
string s = "First thing, "
           "second thing, "
           "third thing";
```
 is so much better than 
```cs
string s = "First thing, " +
           "second thing, " +
           "third thing";
```

What about?
```cs
string myString = "second thing, ";
string s = "First thing, "
           myString
           "third thing";
```
Is that valid?

**Reply by sirinath** [reply]
I am not sure if this should be valid. If this makes this feature consistent and systematic there is not harm in making it valid.

```csharp
string myString = "second thing, ";
string s = "First thing, "
           myString
           "third thing";
```

The 1st 2 cases should be valid.

The 1st and second cases none is better than the other. One can use what ever is clearer or what one prefers based on the content of the string. My preference to drop the `+` is when:
- expressions / code
- has special delimiters
- very short string as. E.g. `"F," +` vs `"F,"` adds 50% more typing and can also be significant when string are 1 - 5 characters.

**Reply by spydacarnage** [reply]
Okay, so as you've intimated that should be valid, this should be, too?
```cs
string This = "This ";
string Is = "is ";
string Valid = "valid ";
string Code = "code!"

string s = This
           Is
           Valid
           Code;
```
And if so, because C# isn't whitespace-bound, this is also valid:
```cs
string s = This Is Valid Code;
```
Can you not see problems with that?

**Reply by jnm2** [reply]
> JS, SQL is one.

@sirinath JS and SQL are better addressed with [raw string literals](https://github.com/dotnet/csharplang/issues/4304) which remove indentation and provide a way to never have to escape the contents, no matter what.

```
var fromRawStringLiteral =
    """
    select Name
    from dbo.People
    where Age >= 21;
    """;
```

Results in this string value:
```
select Name
from dbo.People
where Age >= 21;
```

(Side note: raw string literals is one additional area where the language is whitespace-sensitive, differentiating between newlines and other whitespace characters.)

**Reply by CyrusNajmabadi** [reply]
> very short string as. E.g. "F," + vs "F," adds 50% more typing and can also be significant when string are 1 - 5 characters.

I'm not sure why you'd even write in the new form for such short strings.

**Reply by HaloFour** [reply]
```cs
var s1 = "A"+"B"+"C"+"D"+"E"+"F";
var s2 = "A" "B" "C" "D" "E" "F";
```

**Reply by CyrusNajmabadi** [reply]
Or `var S2 = "ABCDEF";` :D

**Reply by sirinath** [reply]
@jnm2  As motioned before raw literals do not give new line  control (new lines introduce new lines in the string), whitespace control and formatting control (indent the string only without effecting the indentation of the enclosing string content).

@CyrusNajmabadi one use is trading symbols which are 1 - 6 characters.

@CyrusNajmabadi best is to keep it consistent with C/C++ treatment.

@HaloFour 

```csharp
var s1 = "A" +
         "B" +
         "C" +
         "D" +
         "E" +
         "F"; // 2 characters extra for each item
var s2 = "A" 
         "B" 
         "C" 
         "D" 
         "E" 
         "F";
```

Ideally this is more useful for lists:

```csharp
var s1 = "A, " +
         "B, " +
         "C, " +
         "D, " +
         "E, " +
         "F"; // 2 characters extra for each item
var s2 = "A, " 
         "B, " 
         "C, " 
         "D, " 
         "E, " 
         "F";
var s3 = "A, B, C, D, E, F"; // If this is a long, unlike in this example, list split into multiple lines is more clearer
```


This is consistent with C/C++ as C# is a language in the same family. This will give the ability to copy and past strings from one to the other. Reformatting strings can be a pain and not very productive, which can introduce careless mistakes. There is no compelling reason to deviate from C/C++ n this case as this gives some conveniences and also can be a matter of preference. Also gives minor convinces in code porting.

This is a small thing to implement and gives small conveniences which can add up for some developers.

Also who don't prefer this can use alternative with `+`.

**Reply by theunrepentantgeek** [reply]
> This will give the ability to copy and past strings from one to the other.

I've never seen a situation where blindly copying code from one language to another was expected to work. 

In fact, given the oft quoted maxim that **at least** one bug is introduced every time code is copy-pasted, the ability to copy code between languages becomes a tremendous trap.

**Reply by theunrepentantgeek** [reply]
> This is a small thing to implement

No language feature - not one - is small to implement. Every one requires rigorous specifications, implementation, testing, tooling support (from multiple vendors!) and maintenance in perpetuity. 

> Also who don't prefer this can use alternative with +.

Everyone would need to learn the syntax because if _anyone_ uses it, _everyone_ needs to be able to read it.

**Reply by CyrusNajmabadi** [reply]
Having worked on the raw string impl, I can guarantee this is not small :-)

**Reply by sirinath** [reply]
The fact that `+` is made optional is not hard to learn. Any way people from C/C++ and similar languages like MQL4/MQL5 will already familiar with the use.

Also this is not copying code (other than string literals). This is just about copying strings, like common configs, queries, messages, expressions, symbol lists, etc. which do not change from system to system for similar applications.

Also IDE coding style can enforce and convert from one to the other.

**Reply by HaloFour** [reply]
Neither reducing keystrokes as much as possible or attempting to emulate the format of other languages is a goal. The C# designers were well aware of how strings worked in C++ 20 years ago when they made the decision that concatenation is an operation that should be explicit, not one resulting from the accidental vicinity of more than one literal.  The use cases described are covered much better by other features already in (or being added to) the language by other forms of string literals which were also intentionally designed to work very differently than their counterparts in C++.

**Comment by jnm2** [comment]
>  As motioned before raw literals do not give new line control (new lines introduce new lines in the string), whitespace control and formatting control (indent the string only without effecting the indentation of the enclosing string content).

@sirinath With JS and SQL, it is far more common to want newlines in the string. Raw literals *do* provide whitespace and formatting control. (Raw literals let you indent the string only without affecting the indentation of the string content.)
